plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,274 @@
+                    if outstanding_downloads > self.max_concurrent_downloads:
+                        saturated = True
+                        incoming_content = incoming_content[i + 1:]  # remove handled content
+                        break
+                else:
+                    incoming_content = []
+
+                if pending:
+                    done, pending = await asyncio.wait(pending, return_when=asyncio.FIRST_COMPLETED)
+                    for gathered_downloaders in done:
+                        results = gathered_downloaders.result()
+                        for download_result in results[:-1]:
+                            content = results[-1]
+                            for declarative_artifact in content.d_artifacts:
+                                if declarative_artifact.artifact._state.adding:
+                                    new_artifact = Artifact(
+                                        **download_result.artifact_attributes,
+                                        file=download_result.path
+                                    )
+                                    declarative_artifact.artifact = new_artifact
+                            pb.done = pb.done + len(content.d_artifacts)
+                            outstanding_downloads = outstanding_downloads - len(content.d_artifacts)
+                            await out_q.put(content)
+                else:
+                    if shutdown:
+                        break
+
+                if outstanding_downloads < self.max_concurrent_downloads:
+                    saturated = False
+        await out_q.put(None)
+
+
+async def artifact_saver(in_q, out_q):
+    """
+    Stages API stage that saves any unsaved :attr:`DeclarativeArtifact.artifact` objects.
+
+    This stage expects :class:`~pulpcore.plugin.stages.DeclarativeContent` units from `in_q` and
+    inspects their associated :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects. Each
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object stores one
+    :class:`~pulpcore.plugin.models.Artifact`.
+
+    Any unsaved :class:`~pulpcore.plugin.models.Artifact` objects are saved. Each
+    :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to `out_q` after all of its
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects have been handled.
+
+    This stage drains all available items from `in_q` and batches everything into one large call to
+    the db for efficiency.
+
+    Args:
+        in_q (:class:`asyncio.Queue`): The queue to receive
+            :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+        out_q (:class:`asyncio.Queue`): The queue to put
+            :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+
+    Returns:
+        The artifact_saver stage as a coroutine to be included in a pipeline.
+    """
+    shutdown = False
+    declarative_content_list = []
```

Comment:
This name of `declarative_content` is a bit confusing.  I seems to be a *batch* of `DeclarativeContent`.  Having the *type* in the name implies that other variables like `content` are not `DeclarativeContent` and less obvious as to what the list is for.

Recommend renaming to: `batch` or something.

Reply:
I can rename to batch in the next push.